### PR TITLE
Users/sathar/vstest find files fix

### PR DESCRIPTION
**Pull request by SatishThareja**
Manual validation:
1. Pattern based on working directory - include, exclude
2. Pattern based outside working directory - include, exclude
3. Pattern without any regex

Added tests
All UTs passing
Test signoff: sathar,tasoni

### CODE REVIEW

File: Tasks/VsTest/task.json [hunk 1]
```diff
@@ -209,7 +222,7 @@
             "groupName": "reportingOptions"
         }
     ],
-    "instanceNameFormat": "Test Assemblies $(testAssembly)",
+    "instanceNameFormat": "Test Assemblies",
```
Comment (SatishThareja) [inline Tasks/VsTest/task.json]: Need to close on this

File: Tasks/VsTest/task.json [hunk 2]
```diff
@@ -47,11 +48,24 @@
...
         {
             "name": "testAssembly",
-            "type": "string",
+            "type": "multiLine",
             "label": "Test Assembly",
-            "defaultValue": "**\\*test*.dll;-:**\\obj\\**",
+            "defaultValue": "**\\*test*.dll\n!**\\obj\\**",
             "required": true,
-            "helpMarkDown": "Test binaries to run tests on.  Wildcards can be used.  For example, `**\\*test*.dll;-:**\\obj\\**` for all dlls with test in name while excluding files in any sub-directory named obj.",
+            "helpMarkDown": "Test binaries to run tests on. Supports multiple lines of minimatch patterns.",
```
Comment (nigurr) [inline Tasks/VsTest/task.json]: what are the minimatch patterns? ref link to customers? They don't know what's minimatch.
It should be clear in help description how to include how to exclude

File: Tasks/VsTest/task.json [hunk 3]
```diff
@@ -209,14 +223,14 @@
...
-    "instanceNameFormat": "Test Assemblies $(testAssembly)",
+    "instanceNameFormat": "Test Assemblies",
     "execution": {
         "PowerShell": {
             "target": "$(currentDirectory)\\VSTest.ps1",
             "argumentFormat": "",
             "workingDirectory": "$(currentDirectory)",
             "platforms": [
-                "windows"
+                "windows1"
```
Comment (nigurr) [inline Tasks/VsTest/task.json]: what's `windows1`?

File: Tasks/VsTest/task.json [hunk 4]
```diff
@@ -249,6 +263,7 @@
         "ErrorReadingVstestVersion": "Error reading the version of vstest.console.exe.",
         "UnexpectedVersionString": "Unexpected version string detected for vstest.console.exe: %s.",
         "UnexpectedVersionNumber": "Unexpected version number detected for vstest.console.exe: %s.",
-        "VstestDiagNotSupported": "vstest.console.exe version does not support the /diag flag. Enable diagnositics via the exe.config files"
+        "VstestDiagNotSupported": "vstest.console.exe version does not support the /diag flag. Enable diagnositics via the exe.config files",
+        "NoIncludePatternFound": "No include pattern found to search test assemblies."
```
Comment (nigurr) [inline Tasks/VsTest/task.json]: better message?

File: Tasks/VsTest/vstest.ts [hunk 5]
```diff
@@ -24,7 +23,7 @@ try {
     var vsTestVersion: string = tl.getInput('vsTestVersion');
     var vstestLocationMethod: string = tl.getInput('vstestLocationMethod');
     var vstestLocation: string = tl.getPathInput('vsTestLocation');
-    var testAssembly: string = tl.getInput('testAssembly', true);
+    var testAssembly: string[] = tl.getDelimitedInput('testAssembly', '\n', true);
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: how does this work if the input is `a\n\b.dll`?

File: Tasks/VsTest/vstest.ts [hunk 6]
```diff
@@ -51,23 +52,22 @@ try {
...
     }
-    
+
     var releaseuri = tl.getVariable("release.releaseUri")
-	var context = "CI";
-	if(releaseuri)
-	{
-		context = "CD";
-	}
+    var context = "CI";
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: use `let`.

File: Tasks/VsTest/vstest.ts [hunk 7]
```diff
@@ -37,8 +36,10 @@ try {
     var publishRunAttachments: string = tl.getInput('publishRunAttachments');
     var runInParallel: boolean = tl.getBoolInput('runInParallel');
     var tiaEnabled: boolean = tl.getBoolInput('runOnlyImpactedTests');
-    var fileLevel = tl.getVariable('tia.filelevel');
     var tiaRebaseLimit: string = tl.getInput('runAllTestsAfterXBuilds');
+    var searchDirectory: string = tl.getInput('searchDirectory');
+
+    var fileLevel = tl.getVariable('tia.filelevel');
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: let

File: Tasks/VsTest/vstest.ts [hunk 8]
```diff
@@ -51,23 +52,22 @@ try {
...
-		context = "CD";
-	}
+    var context = "CI";
+    if (releaseuri) {
+        context = "CD";
+    }
 
     var systemDefaultWorkingDirectory = tl.getVariable('System.DefaultWorkingDirectory');
     var artifactsDirectory = tl.getVariable('System.ArtifactsDirectory');
     var testAssemblyFiles = getTestAssemblies();
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: stop using global variables across functions. Only entry function should hold of these variables. pass them to functions

File: Tasks/VsTest/vstest.ts [hunk 9]
```diff
@@ -51,23 +52,22 @@ try {
...
 
-    if (testAssemblyFiles && testAssemblyFiles.size != 0) {
-        var workingDirectory = sourcesDir && sourcesDir != '' ? systemDefaultWorkingDirectory : artifactsDirectory;
-        getTestResultsDirectory(runSettingsFile, path.join(workingDirectory, 'TestResults')).then(function (resultsDirectory) {
+    if (testAssemblyFiles && testAssemblyFiles.length != 0) {
+        var workingDirectory = !isNullOrWhitespace(sourcesDir) ? systemDefaultWorkingDirectory : artifactsDirectory;
+        getTestResultsDirectory(runSettingsFile, path.join(workingDirectory, 'TestResults')).then(function(resultsDirectory) {
```
Comment (nigurr) [inline Tasks/VsTest/vstest.ts]: indention

File: Tasks/VsTest/task.json (hunk 2 above)
Comment (SatishThareja) [inline Tasks/VsTest/task.json]: We need a help page for this, which is not ready yet. So, removed the help link for the time being.

File: Tasks/VsTest/vstest.ts (hunk 6 above)
Comment (SatishThareja) [inline Tasks/VsTest/vstest.ts]: changing to let breaks already existing code

File: Tasks/VsTest/task.json (hunk 3 above)
Comment (SatishThareja) [inline Tasks/VsTest/task.json]: Fixed

File: Tasks/VsTest/task.json [hunk 10]
```diff
@@ -47,11 +48,24 @@
...
+                "resizable": "true"
+            }
+        },
+        {
+            "name": "searchDirectory",
+            "type": "string",
+            "label": "Search Directory",
+            "defaultValue": "",
+            "required": false,
+            "helpMarkDown": "Directory to search for the test assemblies.",
```
Comment (vimegh) [inline Tasks/VsTest/task.json]: >"Directory to search for the test assemblies." [](start = 28, length = 46)

Can we somehow add that if left blank it searches in working directory.

File: Tasks/VsTest/task.json (hunk 2 above)
Comment (vimegh) [inline Tasks/VsTest/task.json]: >helpMarkDown [](start = 13, length = 12)

Also, as the pattern is being changed and new Search criteria is being added, hence vsts-tasks/Tasks/VsTest/README.md needs to be updated

File: Tasks/VsTest/Strings/resources.resjson/en-US/resources.resjson [hunk 11]
```diff
@@ -2,13 +2,15 @@
   "loc.friendlyName": "Visual Studio Test",
   "loc.helpMarkDown": "[More Information](https://go.microsoft.com/fwlink/?LinkId=624539)",
   "loc.description": "Run tests with Visual Studio test runner",
-  "loc.instanceNameFormat": "Test Assemblies $(testAssembly)",
+  "loc.instanceNameFormat": "Test Assemblies",
```
Comment (vimegh) [inline Tasks/VsTest/Strings/resources.resjson/en-US/resources.resjson]: any reason for removing $(testAssembly) from instance name format. If I have multiple vs test in build definition I will need to rename each one

File: Tasks/VsTest/task.json (hunk 2 above)
Comment (SatishThareja) [inline Tasks/VsTest/task.json]: Prachi will be closing on the documentation for the new version of task.

File: Tasks/VsTest/Strings/resources.resjson/en-US/resources.resjson (hunk 11 above)
Comment (SatishThareja) [inline Tasks/VsTest/Strings/resources.resjson/en-US/resources.resjson]: Issue with newline char in the task title. Build team will be fixing this bug, and we will need to update it once the fix is available

File: Tasks/VsTest/vstest.ts (hunk 5 above)
Comment (SatishThareja) [inline Tasks/VsTest/vstest.ts]: \n is being used in other tasks as well. this change is required to have uniformity in input between all tasks.

File: Tasks/VsTest/vstest.ts (hunk 7 above)
Comment (SatishThareja) [inline Tasks/VsTest/vstest.ts]: move the code to segregated variables and task inputs

File: Tasks/VsTest/vstest.ts [hunk 12]
```diff
@@ -100,25 +100,79 @@ catch (error) {
...
-function getFilteredFilesFromPattern(pattern: string, hasQuantifier: boolean, allFiles: string[]): string[] {
-    return hasQuantifier ? getFilteredFiles(pattern, allFiles) : [pattern];
+    for (var j = 0; j < pattern.length && pattern[j] === '!'; j++) {
+        negate = !negate;
+        negateOffset++;
+    }
+    patternInfo.negate = negate;
+    if (negate) {
+        tl.debug('exclude pattern: ' + pattern);
+        patternInfo.pattern = pattern.substring(0, negateOffset) + getResolvedPattern(pattern.substring(negateOffset), searchDirectory);
```
Comment (vimegh) [inline Tasks/VsTest/vstest.ts]: >pattern.substring(0, negateOffset) [](start = 30, length = 34)

opt: you can add single bang here

File: Tasks/VsTest/task.json [hunk 13]
```diff
@@ -46,12 +47,25 @@
...
-            "name": "testAssembly",
-            "type": "string",
+            "name": "testAssemblyVer2",
+            "type": "multiLine",
             "label": "Test Assembly",
-            "defaultValue": "**\\*test*.dll;-:**\\obj\\**",
+            "defaultValue": "**\\*test*.dll\n!**\\obj\\**",
             "required": true,
-            "helpMarkDown": "Test binaries to run tests on.  Wildcards can be used.  For example, `**\\*test*.dll;-:**\\obj\\**` for all dlls with test in name while excluding files in any sub-directory named obj.",
+            "helpMarkDown": "Run tests from the specified files. Supports multiple lines of minimatch patterns. [More Information](https://go.microsoft.com/fwlink/?LinkId=624539)",
```
Comment (SatishThareja) [inline Tasks/VsTest/task.json]: https://go.microsoft.com/fwlink/?LinkId=624539 still points to old documentation. Link should be updated once new documentation available.

File: Tasks/VsTest/task.json (hunk 1 above)
Comment (SatishThareja) [inline Tasks/VsTest/task.json]: Input value removed from title due to an issue with newline character in task title. We need to update this update this, once build team fixes this issue.

File: Tasks/VsTest/Strings/resources.resjson/en-US/resources.resjson (hunk 11 above)
Comment (vimegh) [inline Tasks/VsTest/Strings/resources.resjson/en-US/resources.resjson]: discussed offline: there is a tracking bug for this.

---
In reply to: [87799232](https://github.com/Microsoft/vsts-tasks/pull/3065#discussion_r87799232) [](ancestors = 87799232)

File: Tasks/VsTest/typings/vsts-task-lib.d.ts [hunk 14]
```diff
@@ -545,34 +545,52 @@ declare module 'vsts-task-lib/task' {
 	 * @param     tool     path to tool to exec
 	 * @returns   ToolRunner
 	 */
-	export function createToolRunner(tool: string): trm.ToolRunner;
+	export function tool(tool: string): trm.ToolRunner;
+	export interface MatchOptions {
```
Comment (vimegh) [inline Tasks/VsTest/typings/vsts-task-lib.d.ts]: >export interface MatchOptions { [](start = 1, length = 31)

Add comments to tell what is this being used for, similar to other exports

File: Tasks/VsTest/task.json [hunk 15]
```diff
@@ -3,21 +3,22 @@
     "name": "VSTest",
     "friendlyName": "Visual Studio Test",
     "description": "Run tests with Visual Studio test runner",
-    "helpMarkDown": "[More Information](https://go.microsoft.com/fwlink/?LinkId=624539)",
+    "helpMarkDown": "[More Information](https://go.microsoft.com/fwlink/?LinkId=835764)",
```
Comment (SatishThareja) [inline Tasks/VsTest/task.json]: Added new fwdlink, which still points to old documentation. Prachi to update the fwdlink with documentation for preview version.

File: Tasks/VsTest/package.json [hunk 16]
```diff
@@ -1,27 +1,27 @@
...
-    "dependencies": {
-        "xml2js": "0.4.16",
-        "vsts-task-lib": "^0.8.3",
-        "performance-now": "0.2.0"
-    },
-    "optionalDependencies": {
-        "regedit": "2.2.6"
-    }
+  "name": "vsts-tasks-vstest",
+  "version": "2.0.0",
```
Comment (bryanmacfarlane) [inline Tasks/VsTest/package.json]: I would take this version out.  It's not worth maintaining and matching task.json.  Only the task.json is important since this is never published to an npm repository.

File: Tasks/VsTest/task.json [hunk 17]
```diff
@@ -3,21 +3,22 @@
...
-        "Major": 1,
+        "Major": 2,
         "Minor": 0,
-        "Patch": 81
+        "Patch": 0
     },
     "demands": [
         "vstest"
     ],
+    "preview": "true",
```
Comment (bryanmacfarlane) [inline Tasks/VsTest/task.json]: you also need releaseNotes element which contains markdown.  That field offers a description of what this new major version offers.  It is advertised in the web UI for why this customer should try this out.  I think you would say test impact and minimatch find support.

File: Tasks/VsTest/task.json (hunk 1 above)
Comment (bryanmacfarlane) [inline Tasks/VsTest/task.json]: Yes, we're fixing

File: Tasks/VsTest/vstest.ts [hunk 18]
```diff
@@ -51,23 +53,22 @@ try {
...
 
     var systemDefaultWorkingDirectory = tl.getVariable('System.DefaultWorkingDirectory');
     var artifactsDirectory = tl.getVariable('System.ArtifactsDirectory');
     var testAssemblyFiles = getTestAssemblies();
 
-    if (testAssemblyFiles && testAssemblyFiles.size != 0) {
-        var workingDirectory = sourcesDir && sourcesDir != '' ? systemDefaultWorkingDirectory : artifactsDirectory;
-        getTestResultsDirectory(runSettingsFile, path.join(workingDirectory, 'TestResults')).then(function (resultsDirectory) {
+    if (testAssemblyFiles && testAssemblyFiles.length != 0) {
+        var workingDirectory = !isNullOrWhitespace(sourcesDir) ? systemDefaultWorkingDirectory : artifactsDirectory;
```
Comment (bryanmacfarlane) [inline Tasks/VsTest/vstest.ts]: prefer let over var.

File: Tasks/VsTest/vstest.ts [hunk 19]
```diff
@@ -51,23 +53,22 @@ try {
...
+        getTestResultsDirectory(runSettingsFile, path.join(workingDirectory, 'TestResults')).then(function(resultsDirectory) {
             invokeVSTest(resultsDirectory)
-                .then(function (code) {
+                .then(function(code) {
```
Comment (bryanmacfarlane) [inline Tasks/VsTest/vstest.ts]: use await

File: Tasks/VsTest/vstest.ts [hunk 20]
```diff
@@ -81,7 +82,7 @@ try {
                         throw error;
                     }
                 })
-                .fail(function (err) {
+                .fail(function(err) {
```
Comment (bryanmacfarlane) [inline Tasks/VsTest/vstest.ts]: typescript syntax is () => {

File: Tasks/VsTest/vstest.ts [hunk 21]
```diff
@@ -100,30 +101,19 @@ catch (error) {
...
-
-function getTestAssemblies(): Set<string> {
-    return new Set(ffl.findFiles(testAssembly, false, systemDefaultWorkingDirectory));
+    return tl.findMatch(searchFolder, testAssembly, null, getTaskMatchOption());
 }
 
 function getVsTestVersion(): number[] {
     let vstestLocationEscaped = vstestLocation.replace(/\\/g, "\\\\");
-    let wmicTool = tl.createToolRunner("wmic");
+    let wmicTool = tl.tool("wmic");
```
Comment (bryanmacfarlane) [inline Tasks/VsTest/vstest.ts]: good.  also be aware of other changes here:  https://github.com/Microsoft/vsts-task-lib/blob/master/node/docs/releases.md

specifically some folks miss argString is now arg.line.

File: Tasks/VsTest/vstest.ts [hunk 22]
```diff
@@ -247,34 +237,31 @@ function uploadTestResults(testResultsDirectory: string): Q.Promise<string> {
...
-    if (resultFiles && resultFiles[0])
-    {
+    if (resultFiles && resultFiles[0]) {
         selectortool.arg("/ResultFile:" + resultFiles[0]);
     }
     selectortool.arg("/runidfile:" + runIdFile);
     selectortool.arg("/Context:" + context);
     selectortool.exec()
-        .then(function (code) {
+        .then(function(code) {
```
Comment (bryanmacfarlane) [inline Tasks/VsTest/vstest.ts]: use await

File: Tasks/VsTest/vstest.ts [hunk 23]
```diff
@@ -288,42 +275,39 @@ function generateResponseFile(discoveredTests: string): Q.Promise<string> {
...
     selectortool.arg("/BaseLineFile:" + baseLineBuildIdFile);
     selectortool.arg("/platform:" + platform);
-    selectortool.arg("/configuration:" + configuration);    
-	selectortool.arg("/Context:" + context);
+    selectortool.arg("/configuration:" + configuration);
+    selectortool.arg("/Context:" + context);
 
     selectortool.exec()
-        .then(function (code) {
+        .then(function(code) {
```
Comment (bryanmacfarlane) [inline Tasks/VsTest/vstest.ts]: use await

File: Tasks/VsTest/vstest.ts [hunk 24]
```diff
@@ -341,25 +325,21 @@ function publishCodeChanges(): Q.Promise<string> {
...
 
-    var selectortool = tl.createToolRunner(getTestSelectorLocation());
+    var selectortool = tl.tool(getTestSelectorLocation());
     selectortool.arg("PublishCodeChanges");
     selectortool.arg("/TfsTeamProjectCollection:" + tl.getVariable("System.TeamFoundationCollectionUri"));
     selectortool.arg("/ProjectId:" + tl.getVariable("System.TeamProject"));
 
-    if(context == "CD")
-	{	
+    if (context == "CD") {
```
Comment (bryanmacfarlane) [inline Tasks/VsTest/vstest.ts]: http://stackoverflow.com/questions/523643/difference-between-and-in-javascript

File: Tasks/VsTest/vstest.ts [hunk 25]
```diff
@@ -507,17 +487,17 @@ function runVStest(testResultsDirectory: string, settingsFile: string, vsVersion
     var defer = Q.defer<number>();
     if (isTiaAllowed()) {
         publishCodeChanges()
-            .then(function (status) {
+            .then(function(status) {
```
Comment (bryanmacfarlane) [inline Tasks/VsTest/vstest.ts]: wow - look at this indenting and chaining.  use await

File: Tasks/VsTest/task.json [hunk 26]
```diff
@@ -46,12 +47,25 @@
...
+                "resizable": "true"
+            }
+        },
+        {
+            "name": "searchFolder",
+            "type": "string",
+            "label": "Search Folder",
+            "defaultValue": "",
+            "required": false,
+            "helpMarkDown": "Folder to search for the test assemblies. Defaults to $(System.DefaultWorkingDirectory) for build and $(System.ArtifactsDirectory) for release.",
```
Comment (ericsciple) [inline Tasks/VsTest/task.json]: i think system.defaultworkingdirectory is the same as system.artifactsdirectory for a release

File: Tasks/VsTest/vstest.ts [hunk 27]
```diff
@@ -1533,3 +1490,17 @@ function responseContainsNoTests(filePath: string): Q.Promise<boolean> {
...
 }
+
+function isNullOrWhitespace(input) {
+    if (typeof input === 'undefined' || input == null) {
+        return true;
+    }
+    return input.replace(/\s/g, '').length < 1;
+}
+
+function getTaskMatchOption(): any {
```
Comment (ericsciple) [inline Tasks/VsTest/vstest.ts]: we should drive consistency across tasks: `{ dot: true, nobrace: true, nocase: process.platform == 'win32' }` or leave empty and the lib will apply those defaults

File: Tasks/VsTest/typings/vsts-task-lib.d.ts [hunk 28]
```diff
@@ -1,12 +1,12 @@
 declare module 'vsts-task-lib/taskcommand' {
 	export class TaskCommand {
-	    constructor(command: any, properties: any, message: any);
-	    command: string;
-	    message: string;
-	    properties: {
-	        [key: string]: string;
-	    };
-	    toString(): string;
+		constructor(command: any, properties: any, message: any);
```
Comment (ericsciple) [inline Tasks/VsTest/typings/vsts-task-lib.d.ts]: i dont think the d.ts file needs to be checked in. intellisense will come after npm install.

Comment (bryanmacfarlane) [inline Tasks/VsTest/typings/vsts-task-lib.d.ts]: correct.  You never want to check in the vsts-task-lib.  The task lib carries it with the lib which fixed a bunch of problems with them staying in sync.  I worked with the ts team and that's what they recommended.

File: Tasks/VsTest/vstest.ts (hunk 8 above)
Comment (cltshivash) [inline Tasks/VsTest/vstest.ts]: Shouldn't we be able to use classes in typescript ?

---
In reply to: [87553898](https://github.com/Microsoft/vsts-tasks/pull/3065#discussion_r87553898) [](ancestors = 87553898)

**Comment by nigurr** [comment]
Do squash all your changes. too many commits

File: Tasks/VsTest/task.json (hunk 17 above)
Comment (SatishThareja) [inline Tasks/VsTest/task.json]: Added releaseNotes field.

File: Tasks/VsTest/vstest.ts (hunk 22 above)
Comment (SatishThareja) [inline Tasks/VsTest/vstest.ts]: We will do these changes in a separate PR.

File: Tasks/VsTest/vstest.ts (hunk 23 above)
Comment (SatishThareja) [inline Tasks/VsTest/vstest.ts]: We will do these changes in another PR.

File: Tasks/VsTest/vstest.ts (hunk 25 above)
Comment (SatishThareja) [inline Tasks/VsTest/vstest.ts]: We will do these changes in another PR.

File: Tasks/VsTest/vstest.ts (hunk 27 above)
Comment (SatishThareja) [inline Tasks/VsTest/vstest.ts]: Added the default match options and added matchBase for consistency with previous implementation of the task

File: Tasks/VsTest/task.json [hunk 29]
```diff
@@ -209,18 +224,15 @@
...
             "target": "$(currentDirectory)\\VSTest.ps1",
             "argumentFormat": "",
-            "workingDirectory": "$(currentDirectory)",
-            "platforms": [
-                "windows"
-            ]
-        },
-        "Node": {
-            "target": "vstest.js"
+            "workingDirectory": "$(currentDirectory)"
```
Comment (ericsciple) [inline Tasks/VsTest/task.json]: any reason not to delete the powershell handler?

File: Tasks/VsTest/vstest.ts [hunk 30]
```diff
@@ -100,30 +101,18 @@ catch (error) {
...
-}
-
-function getFilteredFilesFromPattern(pattern: string, hasQuantifier: boolean, allFiles: string[]): string[] {
-    return hasQuantifier ? getFilteredFiles(pattern, allFiles) : [pattern];
-}
-
-function getTestAssemblies(): Set<string> {
-    return new Set(ffl.findFiles(testAssembly, false, systemDefaultWorkingDirectory));
+    tl.debug("Searching for test assemblies in: " + searchFolder);
+    return tl.findMatch(searchFolder, testAssembly, null, getTaskMatchOption());
```
Comment (ericsciple) [inline Tasks/VsTest/vstest.ts]: nit: the verbiage in the UI seems to indicate `searchFolder` is the backstop - e.g. calling `find()` on `searchFolder` and then separately calling `match()`.

however, by calling `findMatch`, `searchFolder` will effectively be the "default root" for any unrooted patterns. so you could potentially end up with files outside `searchFolder`. iirc, this is the behavior that you actually want. so my only feedback here is you may consider tweaking the vebiage in the UI.

File: Tasks/VsTest/task.json (hunk 29 above)
Comment (SatishThareja) [inline Tasks/VsTest/task.json]: kept it as an option for user to switch to powershell by updating task.json. We can remove it when we remove other ps scripts

File: Tasks/VsTest/vstest.ts (hunk 27 above)
Comment (ericsciple) [inline Tasks/VsTest/vstest.ts]: since this is a new major version and the goal is consistency across in-the-box tasks i think it would be better to go with `matchBase:false`.

the behavior of matchBase is achievable by prepending the file name with `**\`, so `matchBase:false` doesn't limit functionality.

File: Tasks/VsTest/task.json [hunk 31]
```diff
@@ -46,12 +48,25 @@
...
-            "helpMarkDown": "Test binaries to run tests on.  Wildcards can be used.  For example, `**\\*test*.dll;-:**\\obj\\**` for all dlls with test in name while excluding files in any sub-directory named obj.",
+            "helpMarkDown": "Run tests from the specified files. Supports multiple lines of minimatch patterns. [More Information](https://go.microsoft.com/fwlink/?LinkId=835764)",
+            "groupName": "executionOptions",
+            "properties": {
+                "rows": "3",
+                "resizable": "true"
+            }
+        },
+        {
+            "name": "searchFolder",
```
Comment (SatishThareja) [inline Tasks/VsTest/task.json]: Kept it below the assembly pattern, as it will be empty in most of the cases. 
Used a string to be consistent with other tasks.

File: Tasks/VsTest/vstest.ts (hunk 27 above)
Comment (SatishThareja) [inline Tasks/VsTest/vstest.ts]: okay

File: Tasks/VsTest/vstest.ts (hunk 30 above)
Comment (SatishThareja) [inline Tasks/VsTest/vstest.ts]: updated the help text for assembly pattern for the same.

File: Tasks/VsTest/task.json [hunk 32]
```diff
@@ -46,12 +48,25 @@
     ],
     "inputs": [
         {
-            "name": "testAssembly",
```
Comment (acesiddhu) [inline Tasks/VsTest/task.json]: check the template behaviior. the values given in template will work with V1 task only.

**Comment by acesiddhu** [comment]
:shipit:
